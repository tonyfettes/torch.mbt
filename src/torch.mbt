pub(all) typealias Tensor = @tensor.Tensor

pub fn tensor[X : @tensor.ToTensor](
  x : X,
  requires_grad~ : Bool = false
) -> Tensor {
  tensor(x, requires_grad~)
}

/// TODO: This should be moved into
/// @alert deprecated "Use `distributions` package to initialize an random tensor."
pub fn randn[Shape : @shape.Shape](
  shape : Shape,
  requires_grad~ : Bool = false
) -> Tensor {
  let sample = @distributions.standard_normal.sample(shape)
  if requires_grad {
    sample.requires_grad_(true)
  }
  sample
}
