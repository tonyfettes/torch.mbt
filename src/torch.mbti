package tonyfettes/torch

alias @tonyfettes/torch/shape as @shape
alias @tonyfettes/torch/tensor as @tensor

// Values
fn randn[Shape : @shape.Shape](Shape, requires_grad~ : Bool = ..) -> @tensor.Tensor[Float] //deprecated

fn tensor[X : @tensor.ToTensor](X, requires_grad~ : Bool = ..) -> @tensor.Tensor[Float]

// Types and methods

// Type aliases
pub typealias Tensor = @tensor.Tensor[Float]

// Traits

