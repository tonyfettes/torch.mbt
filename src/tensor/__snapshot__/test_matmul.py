import torch
def test_matmul():
    a = torch.tensor([[-0.2955702841281891,-0.8177422285079956,-0.8997843861579895,-0.154548779129982,-0.5520054697990417,-0.09482517838478088,0.660905659198761,0.21990250051021576]])
    b = torch.tensor([[0.4313355088233948,0.9582411050796509,0.7228865623474121,-0.5188244581222534,0.2701149582862854,-0.21887879073619843,0.9439179301261902,-0.4507488012313843,-0.5474874377250671,0.41745609045028687,0.39493778347969055,0.997472882270813,-0.9003908038139343,0.18648764491081238,-0.717011570930481,-0.2814409136772156],[-0.024237770587205887,0.23831841349601746,0.49988412857055664,-0.2507270872592926,0.4669255316257477,-0.2615886926651001,-0.43666160106658936,-0.15292899310588837,-0.7877539396286011,0.4956502616405487,-0.5569111704826355,0.5138841867446899,0.841296911239624,-0.5421080589294434,-0.8985316157341003,0.8813014030456543],[0.6578092575073242,0.4623033404350281,-0.5396940112113953,-0.47452419996261597,0.8136246204376221,0.40152841806411743,-0.9881891012191772,0.7672386169433594,0.3038009703159332,0.4578361511230469,-0.3085193634033203,-0.24936839938163757,0.5891993641853333,-0.7577323317527771,0.6812232136726379,-0.9833030104637146],[-0.8157637119293213,0.7412014603614807,0.6780391335487366,-0.814010739326477,-0.7987383008003235,-0.42950761318206787,0.9249464273452759,0.5993025898933411,0.5422641038894653,-0.16535788774490356,-0.5273228883743286,-0.06592397391796112,-0.10099833458662033,-0.20775368809700012,-0.3207924962043762,-0.1189662367105484],[-0.9848020076751709,-0.38784119486808777,0.6635101437568665,0.8932644128799438,-0.8225705027580261,-0.14647971093654633,0.9220204949378967,0.37906718254089355,0.16411185264587402,-0.0546141155064106,-0.3278481364250183,-0.26007774472236633,0.5237704515457153,0.15819066762924194,0.784695565700531,0.1746712177991867],[-0.9811980128288269,0.8407623171806335,-0.7188071012496948,-0.8642661571502686,0.939068615436554,0.9759870171546936,-0.7226056456565857,-0.3567482829093933,0.5033784508705139,0.5915883183479309,0.1668814718723297,-0.2744040787220001,0.030391035601496696,0.6279025673866272,0.3989308774471283,-0.6781611442565918],[0.3247348666191101,0.9541763067245483,-0.4985625147819519,0.8650431036949158,-0.3783159852027893,0.40431860089302063,0.4027813971042633,0.8798936605453491,0.4876202642917633,0.4455273449420929,0.6076212525367737,0.9715057611465454,0.29230356216430664,-0.6545597314834595,-0.18756628036499023,-0.10851645469665527],[-0.6491999626159668,0.16038252413272858,-0.9243364334106445,0.9195624589920044,-0.13484114408493042,-0.05838204175233841,-0.033405233174562454,0.8109715580940247,-0.847286581993103,0.7178159356117249,-0.20273204147815704,-0.26399287581443787,-0.4611271917819977,-0.6156734824180603,0.4586758613586426,0.6268856525421143]])
    c = torch.matmul(a, b)
    assert torch.allclose(c, torch.tensor([[0.13503584265708923,-0.20838168263435364,-1.0724889039993286,1.2739464044570923,-0.984972357749939,0.22638460993766785,0.6427057981491089,0.059756532311439514,0.44646602869033813,-0.48874881863594055,1.2199243307113647,0.2731238305568695,-1.1366024017333984,0.38723474740982056,-0.1107659563422203,0.2996768355369568]]))
def test_matmul_multi():
    a = torch.tensor([[0.4483414888381958,0.3644642233848572,-0.9419503808021545],[0.24994567036628723,0.9953191876411438,-0.6054648756980896]])
    b = torch.tensor([[0.6203019022941589,0.6848907470703125],[-0.4301600456237793,0.7356255650520325],[0.7850341796875,0.32139524817466736]])
    c = torch.matmul(a, b)
    d = torch.tensor([[0.6747738718986511,0.8503944277763367,0.2769254148006439,-0.5336059927940369],[-0.41154956817626953,-0.9820307493209839,0.9023672342300415,-0.8075411915779114]])
    e = torch.tensor([[-0.5062331557273865,0.6352475881576538],[0.11726388335227966,-0.15892381966114044],[0.38878488540649414,-0.7222588062286377],[0.6675792336463928,-0.8946297764778137]])
    f = torch.matmul(d, e)
    g = torch.matmul(c, f)
    assert torch.allclose(g, torch.tensor([[0.2772476077079773,-0.36231595277786255],[0.29965129494667053,-0.4518129527568817]]))
