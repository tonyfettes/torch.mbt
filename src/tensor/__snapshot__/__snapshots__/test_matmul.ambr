# serializer version: 1
# name: test_matmul
  tensor([[ 4.4703e-08,  1.4901e-08,  0.0000e+00, -1.1921e-07, -5.9605e-08,
            2.9802e-08,  0.0000e+00,  0.0000e+00, -2.9802e-08, -2.9802e-08,
            1.1921e-07,  0.0000e+00,  0.0000e+00,  0.0000e+00,  8.1956e-08,
            5.9605e-08]])
# ---
